### PR TITLE
add save cache model api in fleet& add slots shuffle in dataset module & add metric op to calculate ctr related metrics

### PR DESCRIPTION
**Pull request by yaoxuefeng6**
(1) add save cache model: add save_cache_model api in fleet. 
usage：
save sparse cache table,
 when using fleet, it will save sparse cache table

 Args:
 dirname(str): save path. It can be hdfs/afs path or local path
 main_program(Program): fluid program, default None
 kwargs: use define property, current support following
 mode(int): 0 currently, only have one default mode and dont check this.

 Example:
 >>> fleet.save_cache_model(dirname="/you/path/to/model", mode = 0)

(2)add slots shuffle: add slots shuffle api in dataset to evaluate slots(feature importance)
usage：
def slots_shuffle(self, slots):
     """
     Slots Shuffle
     Slots Shuffle is a shuffle method in slots level, whicih is usually used
     in sparse feature with large scale of instances. To compare the metric, i.e.
     auc with slots shuffle on one or several slots to evaluate the importance level
     of slots(features).
 
     Args:
         slots(set): the set of slots to do slots shuffle.
 
     Examples:
         import paddle.fluid as fluid
         dataset = fluid.DatasetFactory().create_dataset("InMemoryDataset")
         dataset.set_merge_by_lineid()
         #suppose there is a slot 0
         dataset.slots_shuffle({0})
     """
(3) add metric op to return ctr related metrics
    ctr related metric layer
    This function help compute the ctr related metrics: RMSE, MAE, predicted_ctr, q_value.
    To compute the final values of these metrics, we should do following computations using
    total instance number:
    MAE = local_abserr / instance number
    RMSE = sqrt(local_sqrerr / instance number)
    predicted_ctr = local_prob / instance number
    q = local_q / instance number
    Note that if you are doing distribute job, you should all reduce these metrics and instance
    number first

### CODE REVIEW

File: paddle/fluid/framework/data_feed.cc [hunk 1]
```diff
@@ -35,9 +35,50 @@ limitations under the License. */
...
+
+void RecordCandidateList::ReInit() {
+    _mutex.lock();
+    _full = false;
+    _cur_size = 0;
+    _total_size = 0;
+    _mutex.unlock();
+}
+
+void RecordCandidateList::AddAndGet(const Record& record, RecordCandidate& result, std::shared_ptr<FleetWrapper>& fleet_ptr) {
```
Comment (xjqbest) [inline paddle/fluid/framework/data_feed.cc]: no need to pass "fleet_ptr" here ，you can use "auto fleet_ptr = FleetWrapper::GetInstance();"

File: paddle/fluid/framework/data_feed.h [hunk 2]
```diff
@@ -37,6 +37,7 @@ limitations under the License. */
 #include "paddle/fluid/framework/reader.h"
 #include "paddle/fluid/framework/variable.h"
 #include "paddle/fluid/string/string_helper.h"
+#include "paddle/fluid/framework/fleet/fleet_wrapper.h"
```
Comment (xjqbest) [inline paddle/fluid/framework/data_feed.h]: you can remove

File: paddle/fluid/framework/data_set.cc [hunk 3]
```diff
@@ -114,6 +114,13 @@ void DatasetImpl<T>::SetMergeByInsId(
   keep_unmerged_ins_ = keep_unmerged_ins;
 }
 
+template <typename T>
+void DatasetImpl<T>::SetFeaEval(bool fea_eval, int record_candidate_size) {
+   fea_eval_ = fea_eval;
+   _rclist.ReSize(record_candidate_size);
+   std::cout << "SetFeaEval fea eval mode: " << fea_eval << " with record candidate size: " << record_candidate_size << std::endl;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.cc]: use VLOG(3) here

Comment (xjqbest) [inline paddle/fluid/framework/data_set.cc]: use VLOG(3) here

File: paddle/fluid/framework/data_set.cc [hunk 4]
```diff
@@ -648,5 +655,161 @@ void MultiSlotDataset::MergeByInsId() {
...
+    if (cur_channel_ == 0) {
+        for (size_t i = 0; i < multi_output_channel_.size(); ++i) {
+            out_channel_size += multi_output_channel_[i]->Size();        
+        }
+    } else {
+        for (size_t i = 0; i < multi_consume_channel_.size(); ++i) {
+            out_channel_size += multi_consume_channel_[i]->Size();        
+        }
+    }
+    std::cout << "DatasetImpl<T>::SlotsShuffle() begin with input channel size: " << input_channel_->Size() << " output channel size: " << out_channel_size << std::endl;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.cc]: use VLOG(3) here

File: paddle/fluid/framework/data_set.cc [hunk 5]
```diff
@@ -648,5 +655,161 @@ void MultiSlotDataset::MergeByInsId() {
...
+            }    
+        } else {
+            for (size_t i = 0; i < multi_consume_channel_.size(); ++i) {
+                if (!multi_consume_channel_[i]) {
+                    continue;
+                }
+                end_size += multi_consume_channel_[i]->Size();
+            }   
+        }
+    std::cout << "input channel size before check: " << input_channel_->Size() << " out channel size: " << end_size <<  std::endl;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.cc]: use VLOG(3) here

File: paddle/fluid/framework/data_set.cc [hunk 6]
```diff
@@ -648,5 +655,161 @@ void MultiSlotDataset::MergeByInsId() {
...
+    input_channel_->Write(std::move(random_data));
+    random_data.clear();
+    random_data.shrink_to_fit();
+    input_channel_->Close();
+
+    timeline.Pause();
+    VLOG(3) << "DatasetImpl<T>::SlotsShuffle() end"
+            << ", memory data size for slots shuffle=" << input_channel_->Size()
+            << ", cost time=" << timeline.ElapsedSec() << " seconds";
+    std::cout << "DatasetImpl<T>::SlotsShuffle() end"
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.cc]: remove this duplicate log

File: paddle/fluid/framework/data_set.h [hunk 7]
```diff
@@ -94,6 +96,9 @@ class Dataset {
   virtual void LocalShuffle() = 0;
   // global shuffle data
   virtual void GlobalShuffle() = 0;
+  // for slots shuffle
+  virtual void SlotsShuffle(const std::set<uint16_t>& slots_to_replace) = 0;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.h]: why it's always int，what if user use string as a slot name, such as "myslot"

File: paddle/fluid/framework/data_set.h [hunk 8]
```diff
@@ -168,6 +176,8 @@ class DatasetImpl : public Dataset {
   // and when finish reading, we set cur_channel = 1 - cur_channel,
   // so if cur_channel=0, all data are in output_channel, else consume_channel
   int cur_channel_;
+  std::vector<T> original_data_;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.h]: maybe  shuffle_slot_original_data_  to make it clear

File: paddle/fluid/framework/data_set.h [hunk 9]
```diff
@@ -168,6 +176,8 @@ class DatasetImpl : public Dataset {
...
+  RecordCandidateList _rclist;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.h]: the same as above

File: paddle/fluid/framework/data_set.h [hunk 10]
```diff
@@ -184,13 +194,16 @@ class DatasetImpl : public Dataset {
   bool keep_unmerged_ins_;
   int min_merge_size_;
   std::vector<std::string> merge_slots_list_;
+  bool fea_eval_ = false;
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.h]: the same as above

File: paddle/fluid/framework/data_set.h [hunk 11]
```diff
@@ -184,13 +194,16 @@ class DatasetImpl : public Dataset {
...
 };
 
 // use std::vector<MultiSlotType> or Record as data type
 class MultiSlotDataset : public DatasetImpl<Record> {
  public:
   MultiSlotDataset() {}
   virtual void MergeByInsId();
+  virtual void SlotsShuffle(const std::set<uint16_t>& slots_to_replace);
```
Comment (xjqbest) [inline paddle/fluid/framework/data_set.h]: you should consider user pass a string name of slot。 
it's convenient for user to use "var.name", not a int

File: paddle/fluid/framework/trainer_desc.proto [hunk 12]
```diff
@@ -119,4 +119,7 @@ message TableParameter {
   optional int32 emb_dim = 10;
   optional int32 fea_dim = 11;
   optional string label_var_name = 12;
+  //optional bool enable_sparse_table_cache = 13 [default = true];
```
Comment (xjqbest) [inline paddle/fluid/framework/trainer_desc.proto]: remove these if not used

File: python/paddle/fluid/contrib/layers/__init__.py [hunk 13]
```diff
@@ -16,6 +16,7 @@
 
 from . import nn
 from .nn import *
+
```
Comment (xjqbest) [inline python/paddle/fluid/contrib/layers/__init__.py:19]: remove

File: python/paddle/fluid/dataset.py [hunk 14]
```diff
@@ -91,6 +91,50 @@ def set_pipe_command(self, pipe_command):
...
+            .. code-block:: python
+
+            import paddle.fluid as fluid
+            dataset = fluid.DatasetFactory().create_dataset("InMemoryDataset")
+            dataset.set_fea_eval(1000000, True)
+
+        """
+        if fea_eval:
+            self.dataset.set_fea_eval(fea_eval, record_candidate_size)
+        self.fea_eval = True
```
Comment (xjqbest) [inline python/paddle/fluid/dataset.py]: why is self.fea_eval always True

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py [hunk 15]
```diff
@@ -212,6 +212,48 @@ def save_persistables(self, executor, dirname, main_program=None, **kwargs):
...
+        Args:
+            dirname(str): save path. It can be hdfs/afs path or local path
+            main_program(Program): fluid program, default None
+            kwargs: use define property, current support following
+                mode(int): 0 means save all pserver model,
+                           1 means save delta pserver model (save diff),
+                           2 means save xbox base,
+                           3 means save batch model.
+
+        Example:
```
Comment (xjqbest) [inline python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py:226]: use code block

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py [hunk 16]
```diff
@@ -212,6 +212,48 @@ def save_persistables(self, executor, dirname, main_program=None, **kwargs):
...
+
+        self._role_maker._barrier_worker()
+
+        feasign_num = -1
+        if self._role_maker.is_first_worker():
+            feasign_num = self._fleet_ptr.save_cache(0, dirname, mode)
+
+        self._role_maker._barrier_worker()
+        return feasign_num
+
```
Comment (xjqbest) [inline python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py:252]: remove two lines

File: paddle/fluid/framework/data_feed.cc (hunk 1 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_feed.cc]: fixed

File: paddle/fluid/framework/data_feed.h (hunk 2 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_feed.h]: removed

File: paddle/fluid/framework/data_set.cc (hunk 3 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: fixed

Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: fixed

File: paddle/fluid/framework/data_set.cc (hunk 4 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: fixed

File: paddle/fluid/framework/data_set.cc (hunk 5 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: fixed

File: paddle/fluid/framework/data_set.cc (hunk 6 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: fixed

File: paddle/fluid/framework/data_set.h (hunk 7 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: to do

File: paddle/fluid/framework/data_set.h (hunk 8 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: fixed

File: paddle/fluid/framework/data_set.h (hunk 9 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: fixed it add slots_shuffle prefix

File: paddle/fluid/framework/data_set.h (hunk 10 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: fixed it add slots shuffle prefix

File: paddle/fluid/framework/data_set.h (hunk 11 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: to do

File: paddle/fluid/framework/trainer_desc.proto (hunk 12 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/trainer_desc.proto]: fixed it

File: python/paddle/fluid/contrib/layers/__init__.py (hunk 13 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/contrib/layers/__init__.py:19]: removed

File: python/paddle/fluid/dataset.py (hunk 14 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/dataset.py]: fixed it

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py (hunk 15 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py:226]: added

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py (hunk 16 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py:252]: removed

File: paddle/fluid/framework/data_set.h (hunk 11 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: changed to std::string

File: paddle/fluid/framework/data_set.h (hunk 7 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h]: changed to std::string

File: python/paddle/fluid/contrib/layers/metric_op.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+from paddle.fluid.layer_helper import LayerHelper
+from paddle.fluid.initializer import Normal, Constant
+from paddle.fluid.framework import Variable
+from paddle.fluid.param_attr import ParamAttr
+from paddle.fluid.layers import nn
+
+__all__ = ['ctr_metric_bundle']
+
+
+def ctr_metric_bundle(input, label):
```
Comment (xjqbest) [inline python/paddle/fluid/contrib/layers/metric_op.py:30]: you can also accumulate ins num in this function

File: paddle/fluid/framework/trainer_desc.proto [hunk 18]
```diff
@@ -129,4 +129,7 @@ message TableParameter {
   optional int32 emb_dim = 10;
   optional int32 fea_dim = 11;
   optional string label_var_name = 12;
+  optional bool enable_sparse_table_cache = 13 [ default = true ];
```
Comment (xjqbest) [inline paddle/fluid/framework/trainer_desc.proto]: why default true

Comment (xjqbest) [reply paddle/fluid/framework/trainer_desc.proto]: rm

File: python/paddle/fluid/contrib/layers/metric_op.py (hunk 17 above)
Comment (xjqbest) [reply python/paddle/fluid/contrib/layers/metric_op.py:30]: pos ins num and neg ins num

Comment (xjqbest) [reply python/paddle/fluid/contrib/layers/metric_op.py]: pos ins num and neg ins num

**Review by xjqbest — DISMISSED** [review]

**Review by Thunderbrook — DISMISSED** [review]

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/ps_pb2.py [hunk 19]
```diff
@@ -1335,22 +1407,6 @@
             is_extension=False,
             extension_scope=None,
             options=None),
-        _descriptor.FieldDescriptor(
```
Comment (xjqbest) [inline python/paddle/fluid/incubate/fleet/parameter_server/pslib/ps_pb2.py]: why delete this

Comment (yaoxuefeng6) [reply python/paddle/fluid/incubate/fleet/parameter_server/pslib/ps_pb2.py]: will fix

**Review by xjqbest — DISMISSED** [review]

**Review by fuyinno4 — DISMISSED** [review]

**Review by xjqbest — DISMISSED** [review]

**Review by frankwhzhang — APPROVED** [review]

**Review by hutuxian — APPROVED** [review]

**Pull request merged**